modules/asset-manager-impl/src/main/java/org/opencastproject/assetmanager/impl/endpoint/AbstractAssetManagerRestEndpoint.java

```diff
@@ -272,51 +261,42 @@ public Response getMediaPackage(@PathParam("mediaPackageID") final String mediaP
-                           @PathParam("version") final String version,
-                           @HeaderParam("If-None-Match") final String ifNoneMatch) {
-    return handleException(new P1Lazy<Response>() {
-      @Override public Response get1() {
-        if (StringUtils.isNotBlank(ifNoneMatch)) {
-          return Response.notModified().build();
-        }
-        for (final Version v : getAssetManager().toVersion(version)) {
-          for (Asset asset : getAssetManager().getAsset(v, mediaPackageID, mediaPackageElementID)) {
-            final String fileName = mediaPackageElementID
-                .concat(".")
-                .concat(asset.getMimeType().bind(suffix).getOr("unknown"));
-            asset.getMimeType().map(MimeTypeUtil.Fns.toString);
-            // Write the file contents back
-            return ok(asset.getInputStream(),
-                      Option.fromOpt(asset.getMimeType().map(MimeTypeUtil.Fns.toString)),
-                      asset.getSize() > 0
-                          ? Option.some(asset.getSize())
-                          : Option.<Long>none(),
-                      Option.some(fileName));
-          }
-          // none
-          return notFound();
+                           @PathParam("version") final String version) {
+    try {
+      for (final Version v : getAssetManager().toVersion(version)) {
+        for (Asset asset : getAssetManager().getAsset(v, mediaPackageID, mediaPackageElementID)) {
+          final String fileName = mediaPackageElementID
+                  .concat(".")
+                  .concat(asset.getMimeType().bind(suffix).getOr("unknown"));
+          asset.getMimeType().map(MimeTypeUtil.Fns.toString);
+          // Write the file contents back
+          Option<Long> length = asset.getSize() > 0 ? Option.some(asset.getSize()) : Option.none();
+          return ok(asset.getInputStream(),
+                  Option.fromOpt(asset.getMimeType().map(MimeTypeUtil.Fns.toString)),
+                  length,
+                  Option.some(fileName));
         }
-        // cannot parse version
-        return badRequest("malformed version");
+        // none
+        return notFound();
       }
-    });
+      // cannot parse version
+      return badRequest("malformed version");
+    } catch (Exception e) {
+      return handleException(e);
+    }
   }
 
   /** Unify exception handling. */
-  public static Response handleException(final P1<Response> p) {
-    try {
-      return p.get1();
-    } catch (Exception e) {
-      logger.debug("Error calling REST method", e);
-      Throwable cause = e;
-      if (e instanceof RuntimeException && e.getCause() != null) {
-        cause = ((RuntimeException)e).getCause();
-      }
-      if (cause instanceof UnauthorizedException) {
-        return forbidden();
-      }
-
-      throw new WebApplicationException(e, Response.Status.INTERNAL_SERVER_ERROR);
+  public static Response handleException(Exception e) {
```

Comment:
This non-functional approach is actually shorter, arguably easier to read and does not require Entwine's functional library